src/model_loader_tinygltf.cpp

```diff
@@ -0,0 +1,291 @@
+#include "model_loader_tinygltf.hpp"
+
+#include "util/log.hpp"
+
+#define TINYGLTF_IMPLEMENTATION
+#define STB_IMAGE_IMPLEMENTATION
+#define STB_IMAGE_WRITE_IMPLEMENTATION
+#include <tiny_gltf.h>
+
+namespace wr
+{
+
+	TinyGLTFModelLoader::TinyGLTFModelLoader()
+	{
+		m_supported_model_formats = { "gltf" };
+	}
+
+	TinyGLTFModelLoader::~TinyGLTFModelLoader()
+	{
+	}
+
+	std::pair<std::vector<DirectX::XMFLOAT3>, std::vector<DirectX::XMFLOAT3>> ComputeTangents(ModelMeshData* mesh_data)
+	{
+		size_t vtxCount = mesh_data->m_positions.size();
+		std::vector<DirectX::XMFLOAT3> tanA(vtxCount, { 0, 0, 0 });
+		std::vector<DirectX::XMFLOAT3> tanB(vtxCount, { 0, 0, 0 });
+
+		if (mesh_data->m_uvw.empty())
+		{
+			return { tanA, tanB };
+		}
+
+		// (1)
+		size_t indexCount = mesh_data->m_indices.size();
+		for (size_t i = 0; i < mesh_data->m_indices.size(); i += 3) {
+			size_t i0 = mesh_data->m_indices[i];
+			size_t i1 = mesh_data->m_indices[i + 1];
+			size_t i2 = mesh_data->m_indices[i + 2];
+
+			auto pos0 = DirectX::XMLoadFloat3(&mesh_data->m_positions[i0]);
+			auto pos1 = DirectX::XMLoadFloat3(&mesh_data->m_positions[i1]);
+			auto pos2 = DirectX::XMLoadFloat3(&mesh_data->m_positions[i2]);
+
+			auto tex0 = DirectX::XMLoadFloat2(&DirectX::XMFLOAT2{ mesh_data->m_uvw[i0].x, mesh_data->m_uvw[i0].y });
+			auto tex1 = DirectX::XMLoadFloat2(&DirectX::XMFLOAT2{ mesh_data->m_uvw[i1].x, mesh_data->m_uvw[i1].y });
+			auto tex2 = DirectX::XMLoadFloat2(&DirectX::XMFLOAT2{ mesh_data->m_uvw[i2].x, mesh_data->m_uvw[i2].y });
+
+			DirectX::XMFLOAT3 edge1, edge2;
+			DirectX::XMStoreFloat3(&edge1, DirectX::XMVectorSubtract(pos1, pos0));
+			DirectX::XMStoreFloat3(&edge2, DirectX::XMVectorSubtract(pos2, pos0));
+
+			DirectX::XMFLOAT2 uv1, uv2;
+			DirectX::XMStoreFloat2(&uv1, DirectX::XMVectorSubtract(tex1, tex0));
+			DirectX::XMStoreFloat2(&uv2, DirectX::XMVectorSubtract(tex2, tex0));
+
+			float r = 1.0f / (uv1.x * uv2.y - uv1.y * uv2.x);
+
+			DirectX::XMFLOAT3 tangent(
+				((edge1.x * uv2.y) - (edge2.x * uv1.y)) * r,
+				((edge1.y * uv2.y) - (edge2.y * uv1.y)) * r,
+				((edge1.z * uv2.y) - (edge2.z * uv1.y)) * r
+			);
+
+			DirectX::XMFLOAT3 bitangent(
+				((edge1.x * uv2.x) - (edge2.x * uv1.x)) * r,
+				((edge1.y * uv2.x) - (edge2.y * uv1.x)) * r,
+				((edge1.z * uv2.x) - (edge2.z * uv1.x)) * r
+			);
+
+			tanA[i0] = tangent;
+			tanA[i1] = tangent;
+			tanA[i2] = tangent;
+
+			tanB[i0] = bitangent;
+			tanB[i1] = bitangent;
+			tanB[i2] = bitangent;
+
+		}
+
+		return { tanA, tanB };
+		// (2)
```

Comment:
Unused or debug code should be removed to keep the codebase clean.

Reply:
done